### ISSUE TITLE
Path resolution for in-memory templates

### ISSUE THREAD

**Issue by rendaw** [PebbleTemplates/pebble#542]
StringLoader doesn't provide full implementations for the methods required to use embed/include.  I have templates embedded in a file, where one would expect relative paths to resolve relative to that file.

I think this could be achieved by one of the following:

1. Add a "base path" argument which would be passed to StringLoader and used to resolve relative paths from in-memory templates
2. Make the getTemplate(name, loader) method public so users can create their own loader that does the above

**Comment by nicky9door** [comment]
StringLoader is intended to be used for literal template strings. If you're templates are relative to a file you might want to consider either using FileLoader or otherwise implementing your own Loader class

**Comment by rendaw** [comment]
Right, the issue here is `getTemplate(name, loader)` is private (2 above).  I don't think there's a way to use a custom Loader class ATM.

**Comment by nicky9door** [comment]
You can configure the default Loader in the PebbleEngine.Builder class

**Comment by rendaw** [comment]
Hm, I see.  So I guess it's expected to use the same loader for everything?  And the fact that getLiteralTemplate/getTemplate use different loaders is an unintentional side effect of the api design?

**Comment by nicky9door** [comment]
Yeah. The getLiteralTemplate method was added more recently and internally it was implemented by overriding the loader to be StringLoader

**Comment by rendaw** [comment]
I don't think this will work cleanly
1. Since the loader can't be specified on a per-template basis, all templates have to go through the template loader.  Templates are loaded via a single string so there's no type-safe way to distinguish "in-memory, use custom loader" templates from "filesystem, use fileloader" templates, I'd have to invent some new string-based syntax for distinguishing the two and wouldn't be able to reuse the filesystem loader directly since it wouldn't understand the syntax.
2. The delegating loader appears to use all child loaders for most operations.  If I wanted to make sure for instance the file loader doesn't activate on in-memory templates I'd have to modify it to explicitly check and abort when it finds them.  The file loader behavior isn't documented so I can't assume anything about the internal behavior.
3. The loader that's actually used isn't associated with the template, so I'd need to redo parsing/metadata extraction from the name repeatedly - for instance when resolving relative paths, checking for resource existence, etc.

**Comment by pepperbob** [comment]
Just stumbled over this and I think re-implementing a delegating loader based on string identifiers that follow an URI-schemes shouldn't be too difficult, e.g. referencing templates as `file:/some/file/path`, `db:id-123`. The used schemes would match templates to loaders and it would allow reusing any existing loader.
